pkg/threescale/backend/backend.go

```diff
@@ -633,6 +644,65 @@ func (a *Application) getCacheKey() string {
 	return fmt.Sprintf("%s_%s", a.ownedBy, a.id)
 }
 
+func resetFromIndex(reports []api.UsageReport, index int) []api.UsageReport {
+	for i := index; i < len(reports); i++ {
+		reports[i].CurrentValue = 0
+	}
+	return reports
+}
+
+// handleElapsedTimePeriod iterates over the applications local counters and checks for which known periods have elapsed
+// any period that has already passed the deadline will have its current value counter reset to 0
+func (a *Application) handleElapsedTimePeriod(originalTimestamp int64, deadline time.Time) {
+	a.Lock()
+
+	remoteState := api.UsageReports(a.RemoteState)
+	remoteState.OrderByDescendingGranularity()
+
+	localState := api.UsageReports(a.LocalState)
+	localState.OrderByDescendingGranularity()
+	for metric, reports := range localState {
+		for i, report := range reports {
+			var elapsed bool
+
+			switch {
+			case report.IsForEternity():
+				continue
+			case report.IsForYear():
+				elapsed = time.Unix(originalTimestamp, 0).AddDate(1, 0, 0).After(deadline)
+			case report.IsForMonth():
+				elapsed = time.Unix(originalTimestamp, 0).AddDate(0, 1, 0).After(deadline)
+			case report.IsForWeek():
+				elapsed = time.Unix(originalTimestamp, 0).AddDate(0, 0, 7).After(deadline)
+			case report.IsForDay():
+				elapsed = time.Unix(originalTimestamp, 0).AddDate(0, 0, 1).After(deadline)
+			case report.IsForHour():
+				elapsed = time.Unix(originalTimestamp, 0).Add(time.Hour).After(deadline)
+			case report.IsForMinute():
+				elapsed = time.Unix(originalTimestamp, 0).Add(time.Minute).After(deadline)
+			default:
+				continue
+			}
+			if !elapsed {
+				continue
+			}
+			// we can safely reset the rest without looping through because it is sorted highest to lowest
+			resetFromIndex(reports, i)
```

Comment:
Because of the ordering this should mean that if hour is elapsed, then we also elapse minute, right?

That said, week does not work that way, so if a month is elapsed it might not have an elapsed week, so we must ensure week is treated if present in the remaining reports rather than breaking.

Reply:
ok damn, i thought that ordering them this way would allow us to find the largest period that had elapsed and reset from there. Thats a bit of a nuance then. Will have to rethink